std/algorithm/comparison.d

```diff
@@ -1842,29 +1842,49 @@ store the lowest values.
     assert(min(A(1, "first"), A(1, "second")) == A(1, "first"));
 }
 
-// mismatch
+/*
+Template `canon` is meant as an implementation device only. It is meant to be
+used with the root package of the standard library, as in `canon!"std"`,
+`canon!"std.v2"`, `canon!"std.v3"` etc. Inside the template, implementations
+get to make use of the primitives defined by the selected canon.
+*/
+template canon(string v) {
+    // @@@TODO@@@: this is clowny, must add language support
+    mixin("import "~v~".meta : allSatisfy;");
+    mixin("import "~v~".range.primitives : isInputRange;");
+    mixin("import "~v~".range.primitives : empty, front, popFront;");
+
+    // mismatch
+    static Tuple!Ranges
+    mismatch(alias pred = (a, b) => a == b, Ranges...)(Ranges rs)
+    if (rs.length >= 2 && allSatisfy!(isInputRange, Ranges))
+    {
+        loop: for (; !rs[0].empty; rs[0].popFront)
+        {
+            static foreach (r; rs[1 .. $])
+            {
+                if (r.empty || !binaryFun!pred(rs[0].front, r.front))
```

Comment:
`binaryFun` with a string predicate imports a bunch of `std` stuff (including `std.range`). You need to migrate that as well.

Honestly, ANY functions that you use anywhere should be v2, which should then publicly import the v1 versions if unchanged. Otherwise, any time you want to change a function only for v2, you have to go through every possible usage of it anywhere and upgrade the machinery.

Note that this is kind of implied anyway, because a std.v2 of phobos is going to need to be just as usable as v1 of phobos. But the fact that this worked and pulled in definitions from v1 without explicitly requesting it means this mechanism is prone to subtle v1/v2 mismatch (no pun intended) errors.

I almost think the `canon` template should be in its own file that isn't accidentally importing v1 stuff.

Reply:
Thanks for taking a look!

> `binaryFun` with a string predicate imports a bunch of `std` stuff (including `std.range`). You need to migrate that as well.

Interesting. I'm not so sure. If `std.functional.binaryFun` version 1 works as needed to implement v2's `mismatch` then `mismatch` is free to just use it without migrating it at all. It's a trivial case of code reuse.

> Honestly, ANY functions that you use anywhere should be v2, which should then publicly import the v1 versions if unchanged. Otherwise, any time you want to change a function only for v2, you have to go through every possible usage of it anywhere and upgrade the machinery.

I'm unclear on why this is necessary. Clearly in the long run all symbols are migrated, but during the alpha version of v2 implementations therein should be free to use parts of v1 privately wherever appropriate.

> Note that this is kind of implied anyway, because a std.v2 of phobos is going to need to be just as usable as v1 of phobos. But the fact that this worked and pulled in definitions from v1 without explicitly requesting it means this mechanism is prone to subtle v1/v2 mismatch (no pun intended) errors.

I think this is a liability only if said v1 artifacts are misused, i.e. not used according to their spec.

> I almost think the `canon` template should be in its own file that isn't accidentally importing v1 stuff.

That would probably be a nice option in the future.

Reply:
> Interesting. I'm not so sure. If `std.functional.binaryFun` version 1 works as needed to implement v2's `mismatch` then `mismatch` is free to just use it without migrating it at all. It's a trivial case of code reuse.

It's not a trivial problem. It will result in subtle WTF moments e.g. an alternate form of your unittest:

```d
auto s1 = ["öabc"], s2 = ["üabc"];
auto a = mismatch!((a, b) => a.front == b.front);
assert(a[0].empty && a[1].empty); // all start with the same byte
auto b = mismatch!("a.front == b.front");
assert(a[0] == s1 && a[1] == s2); // but if using string lambdas, autodecoding is back!
```

For `mismatch`, this is probably not going to show up very often, because it has to be a range of ranges AND you must use range functions/algorithms inside your lambda AND it must be a string lambda. But I heavily dislike the subtle inclusion of v1 stuff in corner cases.

Ways to fix:

1. Fix the string lambda mechanism of `binaryFun` so it doesn't import ANY niceties if in v2 mode. You want string lambdas? They have to be simple.
2. Remove string lambda support completely from v2. stirng lambdas are kind of obsolete anyway
3. Fix the code so `binaryFun` imports v2 versions of all those libraries (and fix all those libraries not to use autodecode). That makes this PR significantly bigger and more complex.

> I think this is a liability only if said v1 artifacts are misused, i.e. not used according to their spec.

I think you are headed for a lot of pain if you don't ensure all dependencies go through a v2 gate, even if it's an alias to the v1 version. You did this with `allSatisfy` and `isInputRange`, I'm not sure why you would leave out `binaryFun`.

Reply:
Further note, `tuple` and `Tuple` should be v2 as well.

Reply:
Excellent point, thanks. Incidentally I thought (before realizing this issue) that we should repudiate string lambdas as we have reasonably terse and much more principled lambdas now.

It looks like transitive dependencies must be carefully analyzed to figure out what's what.

Any issues you foresee with `Tuple` and `tuple` related to unexpected incompatibilities?

Reply:
Is stdv2.typecons going to exist? If it is, then either it has the same stuff or modified stuff. Having to know this for reviews, usage, etc. is just a pain. I'd rather just everything that is v2-focused import v2 stuff ONLY, unless it is exposing v1 stuff explicitly (i.e. publicly importing symbols, or using the v1 `canon` template).

If stdv2.typecons is *not* going to exist, then it's not ambiguous, and possibly OK.